### PR TITLE
Fixes a panic in the DNS resolver

### PR DESCRIPTION
**Pull request by SamWhited**
See moby/moby#40715

### CODE REVIEW

File: resolver.go [hunk 1]
```diff
@@ -484,7 +484,7 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			resp, err = co.ReadMsg()
 			// Truncated DNS replies should be sent to the client so that the
 			// client can retry over TCP
-			if err != nil && !resp.Truncated {
+			if err != nil && (resp != nil && !resp.Truncated) {
```
Comment (thaJeztah) [inline resolver.go]: Should we move https://github.com/moby/libnetwork/pull/2540/files#diff-6756fdfda338c13b910a6416f2ac9f54R494-R497 before this instead?

Comment (thaJeztah) [reply resolver.go]: (this code):

```go
			if resp == nil {
				logrus.Debugf("[resolver] external DNS %s:%s returned empty response for %q", proto, extDNS.IPStr, name)
				break
			}
```

Comment (thaJeztah) [reply resolver.go]: oh, I guess not, because then we wouldn't have the `err`

Comment (SamWhited) [reply resolver.go]: `forwardQueryEnd` still needs to be called if `err == nil && resp == nil`, I think. I was trying to shuffle the logic around as little as possible, but I maybe it's worth rewriting some of this?

Comment (thaJeztah) [reply resolver.go]: Yeah,  it's probably ok for now, but we might want to clean it up a bit in a follow-up

**Comment by SamWhited** [comment]
Tests passing!

**Comment by arkodg** [comment]
LGTM on the fix, can you please add a better description in the commit message @SamWhited

**Comment by SamWhited** [comment]
Added a paragraph explaining what changed.

**Comment by thaJeztah** [comment]
did you re-cherry-pick the 19.03 backport as well?

**Comment by SamWhited** [comment]
Sure did

**Pull request merged**